plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/Inline.hs

```diff
@@ -196,18 +195,23 @@ processTerm = handleTerm <=< traverseOf termSubtypes applyTypeSubstitution where
         -> InlineM tyname name uni fun ann (Term tyname name uni fun ann)
     handleTerm = \case
         v@(Var _ n) -> fromMaybe v <$> substName n
-        Let ann NonRec bs t -> do
-            -- Process bindings, eliminating those which will be inlined unconditionally,
-            -- and accumulating the new substitutions.
-            -- See Note [Removing inlined bindings]
-            -- Note that we don't *remove* the bindings or scope the state, so the state will carry
-            -- over into "sibling" terms. This is fine because we have global uniqueness
-            -- (see Note [Inlining and global uniqueness]), if somewhat wasteful.
-            bs' <- wither (processSingleBinding t) (toList bs)
-            t' <- processTerm t
-            -- Use 'mkLet': we're using lists of bindings rather than NonEmpty since we might
-            -- actually have got rid of all of them!
-            pure $ mkLet ann NonRec bs' t'
+        Let ann NonRec bs t -> case bs of
+            b :| [] -> do
+                -- Process the binding, eliminating it if it will be inlined unconditionally,
+                -- and accumulating the new substitutions.
+                -- See Note [Removing inlined bindings]
+                -- Note that we don't *remove* the binding or scope the state, so the state will
+                -- carry over into "sibling" terms. This is fine because we have global uniqueness
+                -- (see Note [Inlining and global uniqueness]), if somewhat wasteful.
+                b' <- processSingleBinding t b
+                t' <- processTerm t
+                -- Use 'mkLet': which takes a possibly empty list of bindings (rather than
+                -- a non-empty list)
+                pure $ mkLet ann NonRec (maybeToList b') t'
+            -- If it is a multi-let, split it and process the bindings one by one, because
+            -- the bindings after `b` need to be considered to determine whether `b` can
+            -- be unconditionally inlined.
```

Comment:
```suggestion
            -- If it is a multi-let, split it and process the bindings one by one, because
            -- the bindings after `b` need to be considered to determine whether `b` can
            -- be unconditionally inlined. Consider e.g. 
            -- `let !x = <effectful>; !y = <effectful> in x`, we need to look at the binding for 
            -- `y` to know that inlining `x` would change the order of effects.
            -- It's awkward to do this when they are part of the same term, (we would be 
            -- looking at "the let term but only considering all the bindings after x") and 
            -- much easier when they are just separate terms.
```

Reply:
Could be a note